src/protocol/aggregation/mod.rs

```diff
@@ -26,110 +27,124 @@ use crate::{
 
 #[step]
 pub(crate) enum Step {
-    BinaryValidator,
-    UpgradeValueBits,
-    UpgradeBreakdownKeyBits,
+    Validator,
+    ConvertValueBits,
 }
 
 /// Binary-share aggregation protocol.
 ///
 /// # Errors
 /// Propagates errors from multiplications
-pub async fn aggregate<'a, C, SB, F, V, BK>(
+pub async fn aggregate<'a, C, S, SB, F, CV, BK>(
     sh_ctx: C,
-    input_rows: &[AggregateInputRow<V, BK>],
+    input_rows: &[AggregateInputRow<CV, BK>],
 ) -> Result<Vec<Replicated<F>>, Error>
 where
     C: UpgradableContext,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F>
+        + BasicProtocols<C::UpgradedContext<F>, F>
+        + Reshare<C::UpgradedContext<F>, RecordId>
+        + Serializable
+        + DowngradeMalicious<Target = Replicated<F>>
+        + 'static,
     C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = SB>,
     SB: LinearSecretSharing<Gf2>
         + BasicProtocols<C::UpgradedContext<Gf2>, Gf2>
         + DowngradeMalicious<Target = Replicated<Gf2>>
         + 'static,
     F: PrimeField + ExtendableField,
-    V: GaloisField,
+    CV: GaloisField,
     BK: GaloisField,
 {
-    let validator = sh_ctx.narrow(&Step::BinaryValidator).validator();
-    let m_ctx = validator.context();
+    let validator = sh_ctx.narrow(&Step::Validator).validator::<F>();
 
-    let (gf2_value_bits, gf2_breakdown_keys) = (
+    let (gf2_value_bits, _gf2_breakdown_keys) = (
         get_gf2_value_bits(input_rows),
         get_gf2_breakdown_key_bits(input_rows),
     );
-    let (upgraded_gf2_value_bits, upgraded_gf2_breakdown_key_bits) = try_join(
-        m_ctx
-            .narrow(&Step::UpgradeValueBits)
-            .upgrade(gf2_value_bits),
-        m_ctx
-            .narrow(&Step::UpgradeBreakdownKeyBits)
-            .upgrade(gf2_breakdown_keys),
-    )
-    .await?;
-    let binary_shared_values = zip(upgraded_gf2_value_bits, upgraded_gf2_breakdown_key_bits)
-        .map(|(value, breakdown_key)| BinarySharedAggregateInputs::new(value, breakdown_key))
-        .collect::<Vec<_>>();
-
-    secure_aggregation(validator, binary_shared_values).await
+
+    // TODO(taikiy):
+    // 1. slice the buckets into N streams and send them to the aggregation protocol
+    // 2. collect the results and return them
+
+    let output = aggregate_values(validator.context(), gf2_value_bits).await?;
```

Comment:
As @martinthomson points out, we can begin with a simple implementation that matches the IPA Aggregation step. It doesn't scale well with large numbers of breakdown keys but it works like this:

1. Use the bit-decomposed breakdown_key value to compute an "equivalence check" with every possible breakdown (value either 0 or 1). This can be done in less multiplications using the "Sterpinkski Thing" we have already implemented.
2. Multiply the contribution by these "equivalence checks" (one multiplication per breakdown... not awesome scaling in the event there are many breakdown keys.
3. Send these multiplication results to the N streams. Most of the times you'll be sending a sharing of zero.

This is fine for now. We should see if we can come up with a more scalable approach.

Reply:
To be clear - I'm suggesting this is the next PR, not this one.